core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -170,89 +187,251 @@ public boolean inodeIdExists(long id) {
+   * @throws FileDoesNotExistException if the target inode does not exist
    */
-  public InodeFile getInodeFileByPath(AlluxioURI path) throws InvalidPathException,
-      FileDoesNotExistException {
-    TraversalResult traversalResult =
-        traverseToInode(PathUtils.getPathComponents(path.toString()), false);
-    if (!traversalResult.isFound()) {
-      throw new FileDoesNotExistException(ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(path));
-    }
-    Inode inode = traversalResult.getInode();
-    if (!inode.isFile()) {
-      throw new FileDoesNotExistException(ExceptionMessage.PATH_MUST_BE_FILE.getMessage(path));
+  public InodePath lockFullInodePath(long id, LockMode lockMode) throws FileDoesNotExistException {
+    for (;;) {
+      Inode<?> inode = mInodes.getFirstByField(mIdIndex, id);
+      if (inode == null) {
+        throw new FileDoesNotExistException("Inode id " + id + " does not exist.");
+      }
+      // Compute the path given the target inode.
+      StringBuilder builder = new StringBuilder();
+      computePathForInode(inode, builder);
+      AlluxioURI uri = new AlluxioURI(builder.toString());
+
+      try (InodePath inodePath = lockFullInodePath(uri, lockMode)) {
+        if (inodePath.getInode().getId() == id) {
+          return inodePath;
+        }
+        // The path does not end up at the target inode id. Repeat the traversal.
+      } catch (InvalidPathException e) {
+        // ignore and repeat the loop
+        LOG.warn("Inode lookup id {} computed path {} mismatch id. Repeating.", id, uri);
+      }
     }
-    return (InodeFile) inode;
   }
 
   /**
-   * Returns a list of existing inodes on the given path.
+   * Attempts to extend an existing {@link InodePath} to reach the target inode. If the target
+   * inode does not exist, an exception will be thrown.
    *
-   * @param path the path to get the inodes list for
-   * @return the inodes list with the given path
+   * @param inodePath the {@link InodePath} to extend to the target inode
+   * @param lockMode the {@link LockMode} to lock the inodes with
    * @throws InvalidPathException if the path is invalid
+   * @throws FileDoesNotExistException if the target inode does not exist
    */
-  public List<Inode<?>> collectInodes(AlluxioURI path) throws InvalidPathException {
-    TraversalResult traversalResult =
-        traverseToInode(PathUtils.getPathComponents(path.getPath()), false);
-    return traversalResult.getInodes();
+  public void ensureFullInodePath(InodePath inodePath, LockMode lockMode)
+      throws InvalidPathException, FileDoesNotExistException {
+    TraversalResult traversalResult = traverseToInode(inodePath, lockMode);
+    if (!traversalResult.isFound()) {
+      throw new FileDoesNotExistException(
+          ExceptionMessage.PATH_DOES_NOT_EXIST.getMessage(inodePath.getUri()));
+    }
+  }
+
+  private void computePathForInode(Inode<?> inode, StringBuilder builder)
```

Comment:
`computePathForInode` is also called from `getPath` which does not perform this check (and it should?)

add a javadoc for this method, which states that the callers are responsible for validating that the path is valid

on a related note, based on full-text search `lockFullInodePath(id, lockMode)` seems to only be used in tests ... if that's the case, we should move it to the test subtree or update the tests to use `lockFullInodePath(uri, lockMode)` instead